src/azul/plugins/metadata/hca/contributor_matrices.py

```diff
@@ -0,0 +1,165 @@
+                }
+            }
+        }
+
+        >>> f([{'uuid': 'u1', 'version': 'v1', 'name': 'n1', 'strata': 'a=1;b=2\\na=3;b=4'},
+        ...    {'uuid': 'u2', 'version': 'v2', 'name': 'n2', 'strata': 'a=1,5;b=7'}])
+        {
+            "a": {
+                "1": {
+                    "b": {
+                        "2": {
+                            "url": [
+                                "https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/u1?version=v1&catalog=dcp"
+                            ]
+                        },
+                        "7": {
+                            "url": [
+                                "https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/u2?version=v2&catalog=dcp"
+                            ]
+                        }
+                    }
+                },
+                "3": {
+                    "b": {
+                        "4": {
+                            "url": [
+                                "https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/u1?version=v1&catalog=dcp"
+                            ]
+                        }
+                    }
+                },
+                "5": {
+                    "b": {
+                        "7": {
+                            "url": [
+                                "https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/u2?version=v2&catalog=dcp"
+                            ]
+                        }
+                    }
+                }
+            }
+        }
+
+        >>> f([{'uuid': 'u1', 'version': 'v1', 'name': 'n1', 'strata': 'a=1;b=2\\na=1'}])
+        Traceback (most recent call last):
+        ...
+        AssertionError: ['a', 'b']
+        """
+        for key in 'uuid', 'name':
+            assert len(set(file[key] for file in files)) == len(files), files
+
+        files = [
+            {
+                # Each line in the stratification string represents a stratum,
+                # each stratum is a list of points, each point has a dimension
+                # and a list of values. Transform that string into a list of
+                # dictionaries. Each entry in those dictionaries maps the
+                # dimension to a value in that dimension. If dimension in a
+                # stratum has multiple values, the stratum is expanded into
+                # multiple strata, one per value. The strata are identical
+                # except in the dimension that had the multiple values.
+                'strata': list(chain.from_iterable(
+                    map(dict, product(*(
+                        [(dimension, value) for value in values.split(',')]
+                        for dimension, values in (point.split('=') for point in stratum.split(';'))
+                    )))
+                    for stratum in file['strata'].split('\n')
+                )),
+                'url': furl(config.service_endpoint(),
```

Comment:
This cannot live here. The plugin cannot know about the service. We need to parameterize this function with a factory for urls. Look for `file_url` in app.py.